torchao/dtypes/fp6.py

```diff
@@ -0,0 +1,160 @@
+        x = x * 2.0 ** (-127 + 3)
+        bits = x.to(tl.int32, bitcast=True)
+
+        sign = ((bits >> 31) & 0x1) << 5
+        exp_and_man = (bits >> 21) & 0x1F
+        result = sign | exp_and_man
+
+        remainder = bits & 0x1F_FFFF
+        do_round_up = (remainder > 0x10_0000) | ((remainder == 0x10_0000) & ((result & 1) == 1))
+        result = tl.where(do_round_up, result + 1, result)
+        return result.to(tl.uint8)
+
+    @triton.jit
+    def _to_fp6_triton_kernel(in_ptr, out_ptr, n, BLOCK_SIZE: tl.constexpr):
+        offsets = tl.program_id(0) * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+        mask = offsets < n
+
+        # strided memory read. there will be uncoalesced memory access
+        val0 = _triton_fp32_to_fp6(tl.load(in_ptr + offsets * 4, mask))
+        val1 = _triton_fp32_to_fp6(tl.load(in_ptr + offsets * 4 + 1, mask))
+        val2 = _triton_fp32_to_fp6(tl.load(in_ptr + offsets * 4 + 2, mask))
+        val3 = _triton_fp32_to_fp6(tl.load(in_ptr + offsets * 4 + 3, mask))
+
+        # bit packing
+        bits0 = (val0 << 2) | (val1 >> 4)  # 0000 0011
+        bits1 = (val1 << 4) | (val2 >> 2)  # 1111 2222
+        bits2 = (val2 << 6) | (val3);      # 2233 3333
+
+        # strided memory write. there will be uncoalesced memory access
+        tl.store(out_ptr + offsets * 3, bits0, mask)
+        tl.store(out_ptr + offsets * 3 + 1, bits1, mask)
+        tl.store(out_ptr + offsets * 3 + 2, bits2, mask)
+
+    def _to_fp6_triton(tensor: Tensor) -> Tensor:
+        out_shape = tensor.shape[:-1] + (tensor.shape[-1] // 4 * 3,)
+        output = torch.empty(out_shape, device=tensor.device, dtype=torch.uint8)
+
+        n = tensor.numel()
+        grid_size = lambda meta: (triton.cdiv(n, meta["BLOCK_SIZE"] * 4),)
+        _to_fp6_triton_kernel[grid_size](tensor, output, n, BLOCK_SIZE=256)
+
+        return output
+
+else:
+    _to_fp6_triton = None
+
+
+def _to_fp6_pt(tensor: Tensor, no_bit_packing: bool = False) -> Tensor:
+    tensor = tensor.float()
+
+    # correct exponent bias. this also handles subnormal numbers correctly
+    tensor = tensor * 2.0 ** (-127 + 3)
+    bits = tensor.view(torch.int32)
+
+    sign = ((bits >> 31) & 0x1) << 5
+    exp_and_man = (bits >> 21) & 0x1F
+    result = sign | exp_and_man
+
+    # round to nearest even
+    remainder = bits & 0x1F_FFFF  # truncated mantissa bits
+    do_round_up = (remainder > 0x10_0000) | ((remainder == 0x10_0000) & ((result & 1) == 1))
+    result = torch.where(do_round_up, result + 1, result)
+    result = result.to(torch.uint8)
+
+    if no_bit_packing:
+        return result
+
+    # bit packing
+    val0, val1, val2, val3 = result.unflatten(-1, (-1, 4)).unbind(-1)
+    bits0 = (val0 << 2) | (val1 >> 4)  # 0000 0011
+    bits1 = (val1 << 4) | (val2 >> 2)  # 1111 2222
+    bits2 = (val2 << 6) | (val3);      # 2233 3333
+    return torch.stack([bits0, bits1, bits2], dim=-1).flatten(-2)
+
+
+def to_fp6(tensor: Tensor, no_bit_packing: bool = False) -> Tensor:
```

Comment:
nit: thoughts about naming your dtype `float6_e3m2` instead of `fp6`?  This is to be consistent with naming for other PyTorch low precision dtypes such as `float8_e4m3|e5m2` from PyTorch core as well as the upcoming MX dtypes, which include `float6_e3m2` and `float6_e2m3`.

Reply:
I was thinking the same thing too! Will update the name.

Reply:
where can I read more about MX dtypes? This particular FP6 used by FP6-LLM paper does not represent +/-inf and NaN, so not sure if we should signal that in the name somehow too? (like `float8_e4m3fn`)

Reply:
You can check out https://www.opencompute.org/documents/ocp-microscaling-formats-mx-v1-0-spec-final-pdf, page 12 describes the supported float6 flavors.  I plan to add the the mx code in torchao soon.

For the `fn` suffix...I'm planning to follow the OCP spec naming, which does not include naming qualifiers for special value handling, and replace `fp` with `float` to be consistent with other PyTorch dtype names.  I think the `fn` suffix made sense for float8 where different flavors had different special value handling, but none of these sub 8 bit dtypes support special values.

Reply:
Cool! It seems like the FP6 I used here is exactly the same as MX FP6 E3M2 (without the scale - FP6 LLM author use 1 scale per row). Perhaps in the future MX dtype can replace this.